src/libAtomVM/ets_hashtable.c

```diff
@@ -98,15 +138,17 @@ EtsHashtableErrorCode ets_hashtable_insert(struct EtsHashTable *hash_table, term
         while (1) {
             if (term_compare(key, node->key, TermCompareExact, global) == TermEquals) {
                 if (opts & EtsHashtableAllowOverwrite) {
-                    node->entry = entry;
+                    node->entry = new_node->entry;
                     memory_destroy_heap(node->heap, global);
-                    node->heap = heap;
+                    node->heap = new_node->heap;
+                    free(new_node);
```

Comment:
I think we should either swap the node entirely or pass contents of it instead of using it as impromptu container, especially when we do swap it on hash collision.